src/Razor/src/Microsoft.CodeAnalysis.Remote.Razor/RazorAssemblyLoadContext.cs

```diff
@@ -0,0 +1,95 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT license. See License.txt in the project root for license information.
+
+#if NET
+using System;
+using System.Diagnostics;
+using System.IO;
+using System.Linq;
+using System.Reflection;
+using System.Runtime.Loader;
+using System.Threading;
+using Microsoft.CodeAnalysis.ExternalAccess.Razor;
+
+namespace Microsoft.CodeAnalysis.Remote.Razor;
+
+internal sealed class RazorAssemblyLoadContext : AssemblyLoadContext
+{
+    private readonly AssemblyLoadContext? _parent;
+    private readonly string _baseDirectory;
+
+    private Assembly? _razorCompilerAssembly;
+
+    private object _loaderLock = new();
+
+    public static readonly RazorAssemblyLoadContext Instance = new();
+
+    public RazorAssemblyLoadContext()
+        : base(isCollectible: true)
+    {
+        var thisAssembly = GetType().Assembly;
+        _parent = GetLoadContext(thisAssembly);
+        _baseDirectory = Path.GetDirectoryName(thisAssembly.Location) ?? "";
+    }
+
+    protected override Assembly? Load(AssemblyName assemblyName)
+    {
+        var fileName = Path.Combine(_baseDirectory, assemblyName.Name + ".dll");
+        if (File.Exists(fileName))
+        {
+            // when we are asked to load razor.compiler, we first have to see if Roslyn beat us to it.
+            if (IsRazorCompiler(assemblyName))
+            {
+                // Take the loader lock before we even try and install the resolver.
+                // This ensures that if we successfully install the resolver we can't resolve the assembly until it's actually loaded
+                lock (_loaderLock)
+                {
+                    if (RazorAnalyzerAssemblyResolver.TrySetAssemblyResolver(ResolveAssembly, assemblyName))
+                    {
+                        // We were able to install the resolver. Load the assembly and keep a reference to it.
+                        _razorCompilerAssembly = LoadFromAssemblyPath(fileName);
+                        return _razorCompilerAssembly;
+                    }
+                    else
+                    {
+                        // Roslyn won the race, we need to find the compiler assembly it loaded.
+                        while (true)
+                        {
+                            foreach (var alc in AssemblyLoadContext.All)
```

Comment:
Consider what happens if another `AssemblyLoadContext` decides to load razor. This strategy of searching all of the `ALC` could lead us to picking that instance vs. the one that Roslyn loaded. Why not just search the Roslyn ALC directly and grab the Razor DLL from there 

```csharp
var roslynAlc = AssemblyLoadContext.GetLoadContext(typeof(Microsoft.CodeAnalysis.Compilation).Assembly);
```

Reply:
Hmm, but it's not loaded into the Roslyn ALC. It's loaded into a shadow copy ALC that's created for the specific directory. Ideally we would check that the path matches our base directory to know its the same assembly but we can't really do that because the shadow copy causes the path to move...

Reply:
Hmm. What if we load the compiler into a metadata only ALC, then we can read the MVID. Then we can check the MVID against the rest of the ALCs to ensure we're picking the correct one?

Reply:
Ah that's true ... but this should still wokr. Rather than iterating through the loaded assemblies just do the following: 

```csharp
var roslynAlc = AssemblyLoadContext.GetLoadContext(typeof(Microsoft.CodeAnalysis.Compilation).Assembly);
var razorCompilerAssembly = roslynAlc.LoadFromAssemblyName(assemblyName);
```

This will go through all of the shadow copy madness in the roslyn ALC and return the assembly if it's available. If not then can proceed to insert the resolver.

Reply:
Oh, that's a neat trick I hadn't thought about. That's a really nice simplification.